# Importing libraries and dataset

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd

In [3]:
white=pd.read_csv("C://Users//Faith//Desktop//WINE//winequality-white.csv",sep=';')

In [4]:
white.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.0              0.27         0.36            20.7      0.045   
1            6.3              0.30         0.34             1.6      0.049   
2            8.1              0.28         0.40             6.9      0.050   
3            7.2              0.23         0.32             8.5      0.058   
4            7.2              0.23         0.32             8.5      0.058   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 45.0                 170.0   1.0010  3.00       0.45   
1                 14.0                 132.0   0.9940  3.30       0.49   
2                 30.0                  97.0   0.9951  3.26       0.44   
3                 47.0                 186.0   0.9956  3.19       0.40   
4                 47.0                 186.0   0.9956  3.19       0.40   

   alcohol  quality  
0      8.8        6  
1      9.5        6  
2     10.1        6  
3      9.9        6  
4      9.9        6

# Exploratory data analysis

In [9]:
white.columns

Index(['fixed acidity', 'volatile acidity', 'citric acid', 'residual sugar',
       'chlorides', 'free sulfur dioxide', 'total sulfur dioxide', 'density',
       'pH', 'sulphates', 'alcohol', 'quality'],
      dtype='object')

In [10]:
white['quality'].nunique()

7

In [11]:
white['quality'].unique()

array([6, 5, 7, 8, 4, 3, 9], dtype=int64)

In [12]:
white.shape

(4898, 12)

In [22]:
red.describe()

fixed acidity  volatile acidity  citric acid  residual sugar  \
count    1599.000000       1599.000000  1599.000000     1599.000000   
mean        8.319637          0.527821     0.270976        2.538806   
std         1.741096          0.179060     0.194801        1.409928   
min         4.600000          0.120000     0.000000        0.900000   
25%         7.100000          0.390000     0.090000        1.900000   
50%         7.900000          0.520000     0.260000        2.200000   
75%         9.200000          0.640000     0.420000        2.600000   
max        15.900000          1.580000     1.000000       15.500000   

         chlorides  free sulfur dioxide  total sulfur dioxide      density  \
count  1599.000000          1599.000000           1599.000000  1599.000000   
mean      0.087467            15.874922             46.467792     0.996747   
std       0.047065            10.460157             32.895324     0.001887   
min       0.012000             1.000000              6.000000     0.990070   
25%       0.070000             7.000000             22.000000     0.995600   
50%       0.079000            14.000000             38.000000     0.996750   
75%       0.090000            21.000000             62.000000     0.997835   
max       0.611000            72.000000            289.000000     1.003690   

                pH    sulphates      alcohol      quality  
count  1599.000000  1599.000000  1599.000000  1599.000000  
mean      3.311113     0.658149    10.422983     5.636023  
std       0.154386     0.169507     1.065668     0.807569  
min       2.740000     0.330000     8.400000     3.000000  
25%       3.210000     0.550000     9.500000     5.000000  
50%       3.310000     0.620000    10.200000     6.000000  
75%       3.400000     0.730000    11.100000     6.000000  
max       4.010000     2.000000    14.900000     8.000000

In [13]:
white.isna().sum()

fixed acidity           0
volatile acidity        0
citric acid             0
residual sugar          0
chlorides               0
free sulfur dioxide     0
total sulfur dioxide    0
density                 0
pH                      0
sulphates               0
alcohol                 0
quality                 0
dtype: int64

In [14]:
red.dtypes

fixed acidity           float64
volatile acidity        float64
citric acid             float64
residual sugar          float64
chlorides               float64
free sulfur dioxide     float64
total sulfur dioxide    float64
density                 float64
pH                      float64
sulphates               float64
alcohol                 float64
quality                   int64
dtype: object

# Classification Analysis

Assigning X and Y variables

In [15]:
X=white.iloc[ :, :-1 ].values
y=white.iloc[:, 11].values

Splitting X and Y into train and test

In [16]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

In [17]:
X_train,X_test,y_train,y_test=train_test_split(X,y, test_size=0.25,random_state=0)


Logistic regression analysis

In [18]:
from sklearn.linear_model import LogisticRegression

In [19]:
log=LogisticRegression()
log.fit(X_train,y_train)
LR=log.predict(X_test)

In [20]:
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report

In [21]:
print(classification_report(y_test,LR))
print(accuracy_score(y_test,LR))

             precision    recall  f1-score   support

          3       0.00      0.00      0.00         9
          4       0.00      0.00      0.00        57
          5       0.57      0.46      0.51       375
          6       0.48      0.83      0.61       522
          7       0.52      0.07      0.12       223
          8       0.00      0.00      0.00        39

avg / total       0.47      0.50      0.44      1225

0.5044897959183674


C:\Users\Faith\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1135: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)


Decision tree analysis

In [22]:
from sklearn.tree import DecisionTreeClassifier

In [23]:
dtc= DecisionTreeClassifier(criterion="entropy")
dtc.fit(X_train,y_train)
DTC=dtc.predict(X_test)

In [24]:
print(classification_report(y_test,DTC))
accuracy_score(y_test,DTC)

             precision    recall  f1-score   support

          3       0.00      0.00      0.00         9
          4       0.37      0.19      0.25        57
          5       0.61      0.60      0.61       375
          6       0.60      0.63      0.61       522
          7       0.50      0.52      0.51       223
          8       0.43      0.46      0.44        39
          9       0.00      0.00      0.00         0

avg / total       0.56      0.57      0.57      1225



C:\Users\Faith\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1135: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)
C:\Users\Faith\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1137: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples.
  'recall', 'true', average, warn_for)


0.5722448979591837

Random forest analysis

In [25]:
from sklearn.ensemble import RandomForestClassifier

In [26]:
rfc=RandomForestClassifier(n_estimators=20, random_state=0)
rfc.fit(X_train,y_train)
RFC=rfc.predict(X_test)

In [27]:
accuracy_score(y_test,RFC)

0.6563265306122449

K Nearest Neighbor Analysis

In [28]:
from sklearn.neighbors import KNeighborsClassifier

In [29]:
#  To calculate the number of k in KNN, find the square root of n
np.sqrt(len(red))

69.98571282769076

In [32]:
knc=KNeighborsClassifier(n_neighbors=70)
knc.fit(X_train,y_train)
KNC=knc.predict(X_test)

In [33]:
accuracy_score(y_test,KNC)

0.43918367346938775

Support Vector Machine

In [34]:
from sklearn.svm import SVC
svc=SVC(kernel='linear')
svc.fit(X_train,y_train)
SV=svc.predict(X_test)

In [35]:
accuracy_score(y_test, SV)

0.5053061224489795

Naive Bayes Classification

In [36]:
from sklearn.naive_bayes import GaussianNB

In [37]:
gnb=GaussianNB()
gnb.fit(X_train, y_train)
GN=gnb.predict(X_test)

In [38]:
accuracy_score(y_test,GN)

0.4424489795918367

# Accuracy Score for all Classification models 

In [39]:
Logistic_Regression= accuracy_score(y_test,LR)
Decision_Tree= accuracy_score(y_test,DTC)
Random_Forest= accuracy_score(y_test,RFC)
K_Nearest_Neighbor= accuracy_score(y_test,KNC)
Support_Vector_Machine= accuracy_score(y_test, SV)
Naive_Bayes= accuracy_score(y_test,GN)

In [42]:
Results=pd.DataFrame(data=[Logistic_Regression,Decision_Tree,Random_Forest,K_Nearest_Neighbor,Support_Vector_Machine,Naive_Bayes],index=columns)

In [41]:
 columns=('Logistic_Regression','Decision_Tree','Random_Forest','K_Nearest_Neighbor','Support_Vector_Machine','Naive_Bayes')

In [43]:
Final_results=Results.transpose()

In [44]:
Final_results

Logistic_Regression  Decision_Tree  Random_Forest  K_Nearest_Neighbor  \
0              0.50449       0.572245       0.656327            0.439184   

   Support_Vector_Machine  Naive_Bayes  
0                0.505306     0.442449

# Results

The result of the above classification analysis indicates that Random_Forest model is the best model for prediction despite having a not so good acuracy score of 0.6563.

---

# Regression Analysis

In [46]:
white.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.0              0.27         0.36            20.7      0.045   
1            6.3              0.30         0.34             1.6      0.049   
2            8.1              0.28         0.40             6.9      0.050   
3            7.2              0.23         0.32             8.5      0.058   
4            7.2              0.23         0.32             8.5      0.058   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 45.0                 170.0   1.0010  3.00       0.45   
1                 14.0                 132.0   0.9940  3.30       0.49   
2                 30.0                  97.0   0.9951  3.26       0.44   
3                 47.0                 186.0   0.9956  3.19       0.40   
4                 47.0                 186.0   0.9956  3.19       0.40   

   alcohol  quality  
0      8.8        6  
1      9.5        6  
2     10.1        6  
3      9.9        6  
4      9.9        6

In [47]:
# To predict the alcohol content, we would assign all columns to the X variable except quality and alcohol.
# y would contain only the alcohol column since it is what we are trying to predict.  
x=red.iloc[:,:-2]
y=red.iloc[:,10]

In [48]:
from sklearn.preprocessing import StandardScaler
scaler=StandardScaler()
scaler.fit(x)
X=scaler.transform(x)

In [49]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.25, random_state=0)

In [50]:
y=y.reshape(-1,1)

C:\Users\Faith\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: FutureWarning: reshape is deprecated and will raise in a subsequent release. Please use .values.reshape(...) instead
  """Entry point for launching an IPython kernel.


Linear Regression

In [51]:
from sklearn.linear_model import LinearRegression
lr=LinearRegression()
lr.fit(X_train,y_train)
LR=lr.predict(X_test)

In [52]:
from sklearn.metrics import r2_score

In [53]:
r2_score(y_test,LR)

0.7256676417394893

Decison tree

In [54]:
from sklearn.tree import DecisionTreeRegressor

In [55]:
dtr= DecisionTreeRegressor(criterion="mse",max_depth=5)
dtr.fit(X_train,y_train)
DTR=dtr.predict(X_test)

In [56]:
r2_score(y_test,DTR)

0.7549460329603118

Random Forest

In [57]:
from sklearn.ensemble import RandomForestRegressor

In [58]:
rfr=RandomForestRegressor(n_estimators=10, random_state=0)
rfr.fit(X_train,y_train)
RFR=rfr.predict(X_test)

In [59]:
r2_score(y_test,RFR)

0.899461222939613

KNN

In [60]:
from sklearn.neighbors import KNeighborsRegressor

In [61]:
knr=KNeighborsRegressor(n_neighbors=40)
knr.fit(X_train,y_train)
KNR=knr.predict(X_test)

In [62]:
r2_score(y_test,KNR)

0.7101841444224783

Support Vector Machine

In [63]:
from sklearn.svm import SVR


In [64]:
svr=SVR()
svr.fit(X_train,y_train)
SV=svr.predict(X_test)

In [65]:
r2_score(y_test,SV)

0.9092411600701336

# Model Evaluation

In [66]:
Linear_Regression=r2_score(y_test,LR)
Decision_Tree=r2_score(y_test,DTR) 
Random_Forest= r2_score(y_test,RFR)
K_Nearest_Neighbor= r2_score(y_test,KNR)
Support_Vector_Machine=r2_score(y_test,SV) 


In [67]:
 columns=('Linear_Regression','Decision_Tree','Random_Forest','K_Nearest_Neighbor','Support_Vector_Machine')

In [68]:
Result=pd.DataFrame(data=[Linear_Regression,Decision_Tree,Random_Forest,K_Nearest_Neighbor,Support_Vector_Machine],index=columns)

In [69]:
Final_results=Result.transpose()

In [70]:
Final_results

Linear_Regression  Decision_Tree  Random_Forest  K_Nearest_Neighbor  \
0           0.725668       0.754946       0.899461            0.710184   

   Support_Vector_Machine  
0                0.909241

# Result

From our model evaluation result, it can be concluded that the support vector machine model
performed best out of all the regression models used with a r2 score of 0.909241 and 
hence shoud be used for our prediction.